### DISCUSSION TITLE
Post new ideas / requests or features in here!

### DISCUSSION THREAD

**Discussion by SimonDarksideJ** [Unity-UI-Extensions/com.unity.uiextensions#416] [Ideas]
## 🙌 We love new ideas and creativity!

Got a new idea or something you would like to see in one of our existing controls, then post about it here.

Anything from:

* New Controls
* Missing features
* Optional add-ons you would like

We would love to hear from you!

**Comment by GordanRamsay** [comment]
Runtime/Scripts/Layout/CardUI/2D Cards/CardPopup2D.cs
Is dependent on Rigidbody which intern requires us to use Physics built-in package.
Please make it so that this layout is made optional.
Or somehow to remove physics dependency.

It is odd for me to be dependent on physics when my game uses no physics, and I have to be dependent on physics because of a UI module.

**Reply by SimonDarksideJ** [reply]
The original author has stopped supporting this control, so I'll have a look into this.  If you can log an issue, I will get it scheduled in the next release.

**Reply by SimonDarksideJ** [reply]
Looking at this now @GordanRamsay , query, would you still want the falling animation or not?  As the whole purpose of this UI element is to provide physics-enabled cards.

Alternate samples provide a similar look (e.g. the 2D Card Stack example) without the physics

**Reply by GordanRamsay** [reply]
Option 1: The ideal scenario I can think of is to add custom-scripted gravity, and basic collision using react's bounds using the Update method of a MonoBehaviour, this would still maintain all features intact, without needing any more actions than just updating the package.
Option 2: Would be to make the 2D Cards an optional download with the package, which would remove the feature from the original repo, but people can fix it by downloading the optional package. This is also easier to implement and can work as quick fix for the problem i mentioned for this thread.

**Reply by GordanRamsay** [reply]
Ill see if I can make those changes and share a pull request.

**Reply by SimonDarksideJ** [reply]
> Option 1: The ideal scenario I can think of is to add custom-scripted gravity, and basic collision using react's bounds using the Update method of a MonoBehaviour, this would still maintain all features intact, without needing any more actions than just updating the package.

This was essentially what I was investigating, but if you can do a PR that allows different types of "gravity" implementations using a common interface, that would be best.

**Comment by mitay-walle** [comment]
[LayoutGroup] Hexagonal Layout Group. Good thing to include
https://github.com/mpewsey/HexagonalUI

[Graphic] My own realization of UICircle, with gradients, split segments etc 
https://github.com/mitay-walle/com.mitay-walle.ui-circle-segmented

[Graphic] My Grid Raw Image for tetris-like inventory
https://github.com/mitay-walle/com.mitay-walle.grid-raw-image

[Graphic] Sector Custom Graphic, for radial menu (Selection wheel)
https://gist.github.com/mitay-walle/dd2e52e5e883878b91e549a9d40bde11

**Reply by SimonDarksideJ** [reply]
Great stuff, will look to including these with the Unity 6 update

**Reply by SimonDarksideJ** [reply]
Requested permission of the author for the components mentioned, will see if they are ok with it

**Reply by SimonDarksideJ** [reply]
Ahh, I see some of them are yours, I'll start porting those now @mitay-walle :D

**Reply by mitay-walle** [reply]
@SimonDarksideJ  here archived demo images for sector

[Sector_demo_images.zip](https://github.com/user-attachments/files/24245920/Sector_demo_images.zip)

**Reply by SimonDarksideJ** [reply]
Just the missing demo content now, saw a reply to the issue, so I'll check that tomorrow

**Comment by mitay-walle** [comment]
Cross-dependencies problem.

Some units in this framework are self-contained 1 file - its great, download 1 file, drop to project and you can use it
Some are on form of folder - its not comfortable to download, then drop to project, but its fine
Some units dependend on some Utility or Base class

I think it would be better to split this dependencies, duplicate and hide in private or nested classes.

It's more often used as library of independent units, the a monolit framework

**Reply by SimonDarksideJ** [reply]
If (when) Unity finally support 3rd party UPM packages, the intention is to migrate the controls to individual "samples" you can import in your project (although it will initially be a pain).  But so far, there have been no complaints about how the package is structured as you get all the controls in one.

But I'm open to suggestions,

**Reply by mitay-walle** [reply]
But custom package with samples is possible, you need to add this section in package.json
https://docs.unity3d.com/Manual/cus-samples.html
Or i missunderstood something?

**Reply by SimonDarksideJ** [reply]
It is possible on OpenUPM, but not in the Unity Asset store. (yet)

**Comment by mitay-walle** [comment]
Lack of documentation

GitHub support upload gifs in comments, its easiest way to add vizualized docs anywhere. 

There is free programs to create gifs from screen, [ScreenToGif](https://www.screentogif.com/)

I can help with this, if pool request would be proceed

![324217079-93fcd863-e140-4f84-a7c0-42a4fd5edad9](https://github.com/Unity-UI-Extensions/com.unity.uiextensions/assets/5578552/4314ac02-6236-4bce-be7f-1529f8fbd18d)

**Reply by SimonDarksideJ** [reply]
Docs are a pain and something I keep chipping away at with each release, so any help there is appreciated, you can find all the online docs here:
https://unity-ui-extensions.github.io/ / https://github.com/Unity-UI-Extensions/Unity-UI-Extensions.GitHub.io

I regularly use gifs where possible in the documentation, when I can :D 

For example, this is the standard I've tried to set for each control:
https://unity-ui-extensions.github.io/Controls/HorizontalScrollSnap.html

**Reply by mitay-walle** [reply]
> For example, this is the standard I've tried to set for each control:
> https://unity-ui-extensions.github.io/Controls/HorizontalScrollSnap.html

I should mention, that I've told closer to overview docs. And principle 'less is more' 

For me good example about docs is Ui Particle from mob sakai : https://github.com/mob-sakai/ParticleEffectForUGUI

Current overview table is good, but it would be awesome, if name of component there will be spoiler, and under it would be hidden 1 long gif with 2-3 use case and features

But dont get me wrong, im grateful for this project, it save my time at work, and i ve inspired from it to start write Packages for unity ui

**Reply by SimonDarksideJ** [reply]
Documentation has gotten an overhaul for the Unity 6 update!

**Comment by zenhund74** [comment]
Dear Simon
I`m still having problems with the UI_Knob extension. I seem to miss to find the right settings.
All I  want to do is to have a Tempo Slider that goes from Tempo 40 to 210 in two loops. But I experience erratic behaviour like a jump to maximum value after one loop. Could you post me the right settings for this endavour? And is it possible to set the slider with a certain value, when I first start the project?

![Bild_2024-12-06_110934549](https://github.com/user-attachments/assets/356f6b59-54e8-41b5-bd89-8ed00af061dc)

**Reply by SimonDarksideJ** [reply]
Probably best logging an issue with your requirements so I can test with the control

**Comment by zenhund74** [comment]
So, if someone else has problems with the UI_Knob settings, we wrote a new script and I add it here. Also you have a new object for the knob handle, so the background picture doesn`t rotate with the handle. As handle you can use a transparent png file with a round red dot or something alike. Kind regards, Jens. If you have questions you can write to info@bluesummerapp.com.

//Credit BlueSummerApp.com
using UnityEngine.EventSystems;
using UnityEngine.Events;
using Unity.Jobs;
using UnityEngine;

public class UIKnob : MonoBehaviour, IDragHandler, IInitializePotentialDragHandler
{
    public float CurrentValue { get; private set; }

    [Header("Knob Settings")]
    [Min(0f)]
    public float MinValue = 0f; // Minimum value of the knob
    [Min(0f)]
    public float MaxValue = 100f; // Maximum value of the knob
    [Min(0f)]
    public float MaxRotation = 720f; // Maximum rotation of the knob (e.g., 720 for 2 full loops)

    [Header("UI References")]
    public RectTransform KnobHandle; // Knob's RectTransform

    public UnityEvent<float> ValueChanged => _valueChanged;
    [SerializeField] private UnityEvent<float> _valueChanged = new UnityEvent<float>();

    private float _currentAngle = 0f; // Current rotation angle
    private Vector2 _knobCenter;

    private float _startDragAngle = 0f;
    private float _accumulatedRotation = 0f;

    void Start()
    {
        if (KnobHandle == null)
        {
            Debug.LogError("Knob Handle is not assigned.");
        }

        // Calculate the center of the knob in local space
        RectTransform rectTransform = GetComponent<RectTransform>();
        _knobCenter = rectTransform.rect.center;
    }

    public void OnInitializePotentialDrag(PointerEventData eventData)
    {
        Vector2 localMousePosition;
        RectTransformUtility.ScreenPointToLocalPointInRectangle(
            KnobHandle.parent as RectTransform,
            eventData.position,
            eventData.pressEventCamera,
            out localMousePosition
        );

        // Calculate the angle relative to the center of the knob
        _startDragAngle = Mathf.Atan2(localMousePosition.y - _knobCenter.y, localMousePosition.x - _knobCenter.x) * Mathf.Rad2Deg;
    }

    public void OnDrag(PointerEventData eventData)
    {
        Vector2 localMousePosition;
        RectTransformUtility.ScreenPointToLocalPointInRectangle(
            KnobHandle.parent as RectTransform,
            eventData.position,
            eventData.pressEventCamera,
            out localMousePosition
        );

        // Calculate the angle relative to the center of the knob
        float currentDragAngle = Mathf.Atan2(localMousePosition.y - _knobCenter.y, localMousePosition.x - _knobCenter.x) * Mathf.Rad2Deg;

        // Calculate the delta angle
        float deltaAngle = Mathf.DeltaAngle(currentDragAngle, _startDragAngle);
        _startDragAngle = currentDragAngle;

        // Update the accumulated rotation and clamp it
        _accumulatedRotation = Mathf.Clamp(_accumulatedRotation + deltaAngle, 0f, MaxRotation);

        // Rotate the knob
        _currentAngle = _accumulatedRotation % 360f;
        KnobHandle.localEulerAngles = new Vector3(0, 0, -_currentAngle);

        // Interpolate the value based on the rotation
        CurrentValue = Mathf.Lerp(MinValue, MaxValue, _accumulatedRotation / MaxRotation);
        _valueChanged.Invoke(CurrentValue);
    }

    public void SetValueWithoutNotify(float value)
    {
        // Clamp the value within the range
        CurrentValue = Mathf.Clamp(value, MinValue, MaxValue);

        // Calculate the corresponding rotation based on the value
        _accumulatedRotation = Mathf.Lerp(0f, MaxRotation, (CurrentValue - MinValue) / (MaxValue - MinValue));

        // Update the knob's rotation
        _currentAngle = _accumulatedRotation % 360f;
        KnobHandle.localEulerAngles = new Vector3(0, 0, -_currentAngle);
    }
}

**Reply by SimonDarksideJ** [reply]
Thanks @zenhund74 , will look to include this with the Unity 6 update

**Reply by SimonDarksideJ** [reply]
Added `UI_Knob2`
WiIl look to add sample with alternate control

**Comment by mitay-walle** [comment]
My Editor Tools for uGUI:

RectTransform buttons to make quick actions
https://github.com/mitay-walle/com.mitay-walle.rect-transform-editor

<img width="865" height="514" alt="{B13731E7-445A-4FD6-9A42-B5A4D9C994BD}" src="https://github.com/user-attachments/assets/2ff37db8-56d0-433d-b1c1-bacd57234f68" />


here is my one-script solution to preview any selectable state (pressed, highlighted)  in form over SceneVew-overlay

https://github.com/mitay-walle/com.mitay-walle.ui-selectable-preview-overlay
<img width="760" height="831" alt="{03D7C4F9-9F6C-4272-8181-17BA72DBA810}" src="https://github.com/user-attachments/assets/0f5b9b03-0bda-48d6-93cc-4035829e972f" />


here is EditorWindow to preview current UI.EventSystem info about pointers (legacy Input and new InputSystem supported)
https://github.com/mitay-walle/DebugUIEventSystemWindow

<img width="614" height="952" alt="image" src="https://github.com/user-attachments/assets/dd9f9c99-7e55-49d2-98be-67cfa7dec689" />

**Comment by mitay-walle** [comment]
My Framework (shader + scripts) to make configurable blending and uv distortion

https://github.com/mitay-walle/com.mitay-walle.ui-sprite-distortion-framework

<img width="784" height="1114" alt="{0CB6B4AD-97B8-40D0-A4F2-551C19F8DFC4}" src="https://github.com/user-attachments/assets/04186d0f-b827-4496-87b6-3401050cb0e6" />

**Comment by zenhund74** [comment]
Hey Simon,
thank you for your reply. We also developed some advanced UI KNob Script.
Works fine in our environment.
Looking forward to discuss,
Jens.
